api_app/analyzers_manager/file_analyzers/pe_info.py

```diff
@@ -7,15 +7,95 @@
 # forked repository: https://github.com/mlodic/pefile
 
 import logging
+import os
 from datetime import datetime
 
+import lief
 import pefile
+import pyimpfuzzy
+from PIL import Image
 
 from api_app.analyzers_manager.classes import FileAnalyzer
+from api_app.exceptions import AnalyzerRunException
 
 logger = logging.getLogger(__name__)
 
 
+class WinPEhasher:
+    file_path: str
+
+    def __init__(self, file_path: str) -> None:
+        """
+        Init a WinPE instance and load the exe file
+        """
+        self.file_path = file_path
+
+    def dhashicon(self) -> str:
+        """
+        Return Dhashicon of the exe
+        """
+        try:
+            # this code was implemented from
+            # https://github.com/fr0gger/SuperPeHasher/commit/e9b753bf52d4e48dda2da0b7801075be4037a161#diff-2bb2d2d4d25fef20a893a4e93e96c9b8c0b0c6d5791fc14594cc9dd5cbf40b41
+            #
+            # config
+            hash_size = 8
+            # extract icon
+            icon_path = self.file_path + ".ico"
+            binary = lief.parse(self.file_path)
+            if binary is None:
+                # Invalid PE file
+                return "PE_Info.WinPEhasher.dhashicon:invalid_pe_file"
+            # extracting icon and saves in a temp file
+            binres = binary.resources_manager
+            if not binres.has_type(lief.PE.RESOURCE_TYPES.ICON):
+                return "PE_Info.WinPEhasher.dhashicon:no_icon_res"
+            ico = binres.icons
+            ico[0].save(icon_path)
+            # resize
+            exe_icon = Image.open(icon_path)
+            exe_icon = exe_icon.convert("L").resize(
+                (hash_size + 1, hash_size), Image.ANTIALIAS
+            )
+            diff = []
+            for row in range(hash_size):
+                for col in range(hash_size):
+                    left = exe_icon.getpixel((col, row))
+                    right = exe_icon.getpixel((col + 1, row))
+                    diff.append(left > right)
+            decimal_value = 0
+            icon_hash = []
+            for index, value in enumerate(diff):
+                if value:
+                    decimal_value += 2 ** (index % 8)
+                if (index % 8) == 7:
+                    icon_hash.append(hex(decimal_value)[2:].rjust(2, "0"))
+                    decimal_value = 0
+            os.remove(icon_path)
+            return "".join(icon_hash)
+        except Exception as e:
+            raise AnalyzerRunException(
+                f"pe_info.winpe_hasher.dhashicon gave errors. {str(e)}"
```

Comment:
I noticed now: for both the cases, we should not raise an AnalyzerRunException if they fail. Otherwise we make this analyzer fails for just a component and this does not make sense because the rest of the code could work.

So, in the case of error, we should just log the error as a warning (becasue it is not critical), save the error in the report and return an empty string for this hash.

```
            logger.warning(warning_message)
            self.report.errors.append(warning_message)
```

Reply:
Ok Sir

Reply:
this still raises an `AnalyzerRunException` so why was marked as "solved"?